### PR TITLE
feat: add get-dependent-fils utils

### PR DESCRIPTION
**Pull request by ashoktamang**
utility file contains function getDependentFiles which provides a way for 
files access, AST parse for ModuleSpecifier, map for dependent files.

(Added dependencies: `denodeify` and `es6-promise` in `package.json`)

### CODE REVIEW

File: addon/ng2/utilities/get-dependent-files.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+
+interface MapType {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This seems more like a type, which can be declared as:
```typescript
declare type MapType = {
  [key: string]: MapType | string[];
};
```

I'm also kinda confused as of the role of this type. It's a `string => string[]` mapping with an hierarchical recursion on the type, but I don't see you use any key with `MapType`. Is it just a `string => string[]` mapping?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+/**
+ * Traverse through ts.Node recursively and stores moduleSpecifiers in an array
+ * 
+ * @param {ts.Node} node: Typescript Node of whose AST is being traversed
+ * @param {string[]} moduleSpecifiers: array to store moduleSpecifiers if found.
+ * 
+ */
+export function getModuleSpecifier(node: ts.Node, moduleSpecifiers: any[]): any {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: If possible, don't use `any`.  Your `@param` description seems to indicate this is

File: addon/ng2/utilities/get-dependent-files.ts (hunk 1 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: I wanted to map relevant files with their relative path imports. So, it seems I can omit this declaration, store the results in an object.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 2 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Done.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 1 above)
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: The name is misleading. Maybe using something like `ImportMap` (no need for the type suffix)?

Your keys can be the path of the file, the value an array of relative/absolute paths of imports. What do you think?

Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: I was doing exactly the same thing. So, do you want me to just change `MapType` to `ImportMap` even though this mapping will a `string => string[]` only?

Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: In order to facilitate readability where it's used, the type name should reflect its content (the same reason you don't name your variables with generic names), so yes I think it's best if the type has the name of how it will be used :)

Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Got it. Thanks 👍. Done.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/**
+ * Create a SourceFile as defined by Typescript Compiler API.
+ * Generate a AST structure from a source file.
+ * 
+ * @param fileName source file for which AST is to be extracted
+ */
+export function createtsSourceFile(fileName: string): ts.SourceFile {
+  return ts.createSourceFile (
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Styling issues: this is hard to read. No spaces between the function and `(`, put the first argument on the same line as the call and horizontally align additional arguments with it. Align the `)` at the end of the last argument.

Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Done.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  let relevantFiles: ImportMap = {};
+  files.forEach(file => {
+    let moduleSpecifiers: any[] = [];
+    let tsFile = createtsSourceFile(file);
+    if (moduleSpecifiers) {
+      getModuleSpecifier(tsFile, moduleSpecifiers);
+      allFiles[file] = moduleSpecifiers;
+    }
+  });
+  for (let filePath in allFiles) {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Prefer to use `for (const filePath of Object.keys(allFiles))` to `for (... in ...)`. There are a lot of reasons why `for (... in ...)` won't work the way you expect it to (in some cases).

File: addon/ng2/utilities/get-dependent-files.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      }
+    }
+  });
+
+  let allFiles: ImportMap = {};
+  let relevantFiles: ImportMap = {};
+  files.forEach(file => {
+    let moduleSpecifiers: any[] = [];
+    let tsFile = createtsSourceFile(file);
+    if (moduleSpecifiers) {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: You just defined `moduleSpecifiers`. Is there a case where this `if` would be false?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ *  Store all the files' path under the given path in a passed in array.
+ *  All the files in the subdirectories of the given path are also included.
+ * 
+ * @param filePath path from where all the files' path are retrieved
+ * @param allFiles array where the files are stored and then returned.
+ * 
+ */
+
+export function getFiles(filePath: string, allFiles: string[]): any {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This look hugely like a broken implementation of the `glob` node module. Check it out.  This implementation seems broken because you automatically assume anything without an extension is a directory, which is wrong (directories can have extensions and files can be without extensions).

In any case, I recommend removing this and using `glob`, which is already installed. This function can be simplified to ```typescript
const allFiles = glob.sync(path.join(somePath, '**/*'), { nodir: true });
```.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  [key: string]: ImportMap | string[];
+}
+
+/**
+ * Create a SourceFile as defined by Typescript Compiler API.
+ * Generate a AST structure from a source file.
+ * 
+ * @param fileName source file for which AST is to be extracted
+ */
+export function createtsSourceFile(fileName: string): ts.SourceFile {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: `createTsSourceFile`?  `create` and `ts` are two different words :)

File: addon/ng2/utilities/get-dependent-files.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  return ts.createSourceFile(fileName,
+                             fs.readFileSync(fileName).toString(),
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This could be simplified to `ts.readFileSync(fileName, 'utf-8')`, which returns a JS string.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+/**
+ * Traverse through ts.Node recursively and stores moduleSpecifiers in an array
+ * 
+ * @param {ts.Node} node: Typescript Node of whose AST is being traversed
+ * @param {string[]} moduleSpecifiers: array to store moduleSpecifiers if found.
+ * 
+ */
+export function getModuleSpecifier(node: ts.Node, moduleSpecifiers: string[]): any {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: I notice you create a lot of functions with side effects. This is kind of frowned upon because people using your functions would probably not expect it. There's almost no advantage (surprisingly) of doing it that way rather than returning new objects, so in general we prefer to have pure functions.

To make this function pure, remove the moduleSpecifiers argument and create a local array. Then return that array at the end of the function (like you're already doing). When recursively calling `getModuleSpecifier`, instead do `moduleSpecifiers.push(...moduleSpecifiers(nodeObj));` or similar.

Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Also, any reasons this function is `export`ed?  Seems like no one would like to import it, and if you export it just for testing you can test all code paths in this function from the actual API. What do you thnk?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  node.getChildren().forEach(nodeObj => {
+    if (node.kind === ts.SyntaxKind.ImportClause) {
+      let moduleSpecifier = node.parent.moduleSpecifier;
+      let instance: MapType = {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Is `MapType` still a thing?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * Traverse through ts.Node recursively and stores moduleSpecifiers in an array
+ * 
+ * @param {ts.Node} node: Typescript Node of whose AST is being traversed
+ * @param {string[]} moduleSpecifiers: array to store moduleSpecifiers if found.
+ * 
+ */
+export function getModuleSpecifier(node: ts.Node, moduleSpecifiers: string[]): any {
+  node.getChildren().forEach(nodeObj => {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Here's a cool trick; you can use reduce to make your code more flow-like:

```typescript
declare interface ModuleImport {
  specifier: string,
  pos: number,
  end: number
};

function _getImportClauses(node: ts.Node): ModuleImport[] {
  return node.getChildren()
    .reduce(function(nodes, node) {
      if (current.kind === ts.SyntaxKind.ImportClause) {
        return nodes.concat({ specifier: node.text, pos: node.pos, end: node.end });
      } else {
        return nodes.concat(_getImportClauses(node));
      }
    }, [])
}
```

You don't have to if you don't want to do it, it just reads a bit better but your version isn't bad :)

File: addon/ng2/utilities/get-dependent-files.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+ * Return a map of all dependent file/s' path with their moduleSpecifier object
+ * (text, pos, end)
+ * 
+ * @param fileName file upon which other files depend 
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): MapType {
+  let files: string[] = [];
+  getFiles(rootPath, files);
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Using `glob`, just do `const files = glob.sync(path.join(rootPath, '**/*.component.ts'));` and skip the next 8 or so lines.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import * as fs from 'fs';
+import * as ts from 'typescript';
+import * as glob from 'glob';
+import * as path from 'path';
+
+/**
+ * Declare a MapType where key is a string and value is an array of type string
+ * Use for storing a file's path as key and its dependent module's relative path
+ */
+declare interface ModuleImport {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: You don't need the `declare` keyword here - interfaces are automatically declarations.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  end: number;
+};
+
+/**
+ * Create a SourceFile as defined by Typescript Compiler API.
+ * Generate a AST structure from a source file.
+ * 
+ * @param fileName source file for which AST is to be extracted
+ */
+export function createTsSourceFile(fileName: string): ts.SourceFile {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: This function should probably not be exported, as it's not public functionality of a utility to get dependent files.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  return ts.createSourceFile(fileName,
+                             fs.readFileSync(fileName, 'utf-8'),
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: @hansl, how do you feel about `readFileSync` in the CLI?

My concern would be that doing this - while convenient - precludes optimizing operations like this through concurrency in the future.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+                             ts.ScriptTarget.ES6,
+                             true);
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Don't indent like this - four spaces is enough to signify these parameters belong to the call from the line above.

Also, move `fileName` to the next line when you do this, which makes it more readable.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    .reduce(function(nodes, node) {
+      if (node.kind === ts.SyntaxKind.ImportClause) {
+        let moduleSpecifier = node.parent.moduleSpecifier;
+        return nodes.concat({
+          specifier: moduleSpecifier.text,
+          pos: moduleSpecifier.pos,
+          end: moduleSpecifier.end
+        });
+      } else {
+        return nodes.concat(_getImportClauses(node));
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: I don't think you need this - imports will always be at the top level of the file.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Traverse through ts.Node recursively and stores moduleSpecifiers in an array
+ * 
+ * @param {ts.Node} node: Typescript Node of whose AST is being traversed
+ * @param {string[]} moduleSpecifiers: array to store moduleSpecifiers if found.
+ * 
+ */
+function _getImportClauses(node: ts.Node): ModuleImport[] {
+  return node.getChildren()
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: `node` will be a `SourceFile` so just use `SourceFile.statements` instead of `getChildren()`.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+ * Return a map of all dependent file/s' path with their moduleSpecifier object
+ * (text, pos, end)
+ * 
+ * @param fileName file upon which other files depend 
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): MapType {
+  const files = glob.sync(path.join(rootPath, '**/*.component.ts'), { nodir: true });
+  let allFiles: ImportMap = {};
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: What is `ImportMap`?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Return a map of all dependent file/s' path with their moduleSpecifier object
+ * (text, pos, end)
+ * 
+ * @param fileName file upon which other files depend 
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): MapType {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Should this be `ModuleImport` instead of `MapType`?

File: tests/runner.js [hunk 21]
```diff
@@ -35,7 +35,7 @@ var glob = require('glob');
 var path = require('path');
 
 var root = 'tests/{acceptance,models,e2e}';
-var specFiles = glob.sync(root + '/**/*.spec.*');
+var specFiles = glob.sync(root + '/**/get-dependent-files.spec.*');
```
Comment (alxhub) [inline tests/runner.js]: This change seems like it made it in on accident.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 15 above)
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Yeah `readFile` would be better, and you could return a `Promise<ts.Sourcefile>` instead. This means that you have to return a `Promise<ModuleImport[]>` in `_getImportClauses()`, and wait for all promises to be resolved.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+ * Use for storing a string literal, start position and end posittion of ImportClause node kinds.
+ */
+interface ModuleImport {
+  specifier: string;
+  pos: number;
+  end: number;
+};
+
+export interface ModuleMap {
+  [key: string]: ModuleImport;
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Since this interface is exported, `ModuleImport` should be too or Typescript will complain.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+ * 
+ * @param {ts.Node} node: Typescript Node of whose AST is being traversed
+ * @param {string[]} moduleSpecifiers: array to store moduleSpecifiers if found.
+ * 
+ */
+function _getImportClauses(node: ts.SourceFile): ModuleImport[] {
+  let moduleSpecifiers: ModuleImport[] = [];
+  node.statements.forEach((nodeObj: ts.Node) => {
+    // Every object in the node.statements has an child node object called moduleSpecifier
+    // TODO: investigate on the cases where this assumption is not true.
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: This assumption is _definitely_ not true. Only import statements will have module specifiers.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Return a map of all dependent file/s' path with their moduleSpecifier object
+ * (text, pos, end)
+ * 
+ * @param fileName file upon which other files depend 
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): Promise<ModuleImport> {
+  const files = glob.sync(path.join(rootPath, '**/*.component.ts'), { nodir: true });
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Process all Typescript files, not just components.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+
+  let allPromiseFiles = files
+    .map(file => createTsSourceFile(file))
+    .map(tsFile => tsFile
+      .then((node: ts.Node) => _getImportClauses(node)));
+
+  return Promise
+    .all(allPromiseFiles).then((moduleSpecifiers: ModuleImport[]) => {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Split `.all` and `.then` onto separate lines.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+      if (nodeObj.kind === ts.SyntaxKind.ImportDeclaration) {
+        nodeObj.getChildren().forEach((childNode: ts.Node) => {
+          if (childNode.kind === ts.SyntaxKind.StringLiteral) {
+            let specifierText = childNode.getFullText();
+            /** 
+             * ts.node.getFullText() returns a string with quotes.
+             * Stripping quotes from the string before adding in the object.
+             * Example: "import * from './x.component'",
+             * if the return value from ts.node.getFullText() is stored in moduleSpecifiers, 
+             * then the string is stored as: '\'./x.component''
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This confuses me a little. Is the first `'` escaped and not the last?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+    });
+}
+
+/**
+ * @return traverses through ts.Node and returns an array of moduleSpecifiers
+ * 
+ * @param {ts.SourceFile} node: Typescript Node of whose AST is being traversed
+ * 
+ */
+export function getImportClauses(node: ts.SourceFile): ModuleImport[] {
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This function lacks unit tests.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 26 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: I made a mistake there. It is escaped both at the first and last `'`. For instance, the function: `node.getFullText()` returns: 
`' \'./x.component\''`

File: addon/ng2/utilities/get-dependent-files.ts (hunk 27 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: My reasoning for not having unit tests on this function is that `getDependentFiles(arg, arg)` uses this function.

But since I am also using this function for other functions that rewrites imports, I should have tests, right?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+  let moduleSpecifiers: ModuleImport[] = [];
+  if (node.statements.length > 0) {
+    node.statements.forEach((nodeObj: ts.Node) => {
+      if (nodeObj.kind === ts.SyntaxKind.ImportDeclaration) {
+        nodeObj.getChildren().forEach((childNode: ts.Node) => {
+          if (childNode.kind === ts.SyntaxKind.StringLiteral) {
+            let specifierText = childNode.getFullText();
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: You can compress the lines above with the following (and gain a bit of readability):

```typescript
return node.statements
  .filter(node => node.kind === ts.SyntaxKind.ImportDeclaration)  // Only Imports.
  .map((node: ts.ImportDeclaration) => ({
    specifierText: node.moduleSpecifier.slice(1, -1),
    pos: node.pos,
    end: node.end
  }));
```

This uses the fact that an [ImportDeclaration](https://github.com/Microsoft/TypeScript/blob/a03941237334f6757c932254d58f1833fa1b004a/lib/tsserverlibrary.d.ts#L926) already has a member with what you're looking for :)

Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Wow, this is very neat and comprehensible. Thanks, I will use this.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 27 above)
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: 2 things about unit tests; coverage and examples.

Coverage; unless you can use `getDependentFiles` to cover every possible branches and statements in this function, you should also make unit tests for it.

Example; unit tests can be used by people reading your code to better understand how to use it / how it works. If I see what goes in and what comes out, it's easier for me to get the goal of this function. Basically, tests are also part of the documentation.

Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: understood. 👍

File: addon/ng2/utilities/get-dependent-files.ts (hunk 28 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Btw, `node.moduleSpecifier.text` provides the desired specifierText. I was having problem only when using `node.getFullText()` to get the text.

File: tests/acceptance/get-dependent-files.spec.ts [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+      });
+    });
+    it('when there are no dependent files', () => {
+      let sourceFile = path.join(rootPath, 'foo-baz/no-module.component.ts');
+      return dependentFilesUtils.getDependentFiles(sourceFile, rootPath)
+      .then((contents: dependentFilesUtils.ModuleMap) => {
+        assert.deepEqual(contents, {});
+      });
+    });
+ });
```
Comment (hansl) [inline tests/acceptance/get-dependent-files.spec.ts]: Indentation is wrong for this line.

Comment (ashoktamang) [inline tests/acceptance/get-dependent-files.spec.ts]: Done.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): Promise<ModuleMap> {
+  const files = glob.sync(path.join(rootPath, '**/*.*.ts'), { nodir: true });
+
+  let allPromiseFiles = files
+    .map(file => createTsSourceFile(file))
+    .map(tsFile => tsFile
+      .then((node: ts.SourceFile) => getImportClauses(node)));
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: This is hard to follow because `tsFile` is actually misnamed; it's a promise of a `SourceFile`, really. It's kinda confusing to have a map return a promise and another map wait on that promise. I feel it's more natural to do it the other way.

What you could also do it reverse the logic a little bit:

```typescript
return Promise.all(files.map(file => createTsSourceFile(file)))
  .then(tsFiles => tsFiles.map(file => getImportClauses(file))
  .then((moduleSpecifiers: ModuleImport[][]) => { ... })
  ...
```

So waiting on the createTsSourceFile instead of the getImportClauses return. It's up to you; what do you think?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 31]
```diff
@@ -0,0 +1,132 @@
...
+        // normalize path of the files for tests compatability (POSIX and Windows)
+        let sourcePath = path.normalize(files[index]);
+        allFiles[sourcePath] = moduleSpecifiers[index];
+      }
+      return allFiles;
+    })
+    .then((allFiles: ModuleMap) => {
+      let relevantFiles: ModuleMap = {};
+      Object.keys(allFiles).forEach(filePath => {
+        let tempModuleSpecifiers: ModuleImport[] = [];
```
Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: The logic here seems overly complex as well, but it's understandable. Looking at the code, it seems like `tempModuleSpecifiers` would be a subset of `allFiles` values, am I correct? In that case you could simplify by using a filter `const tempModuleSpecifiers = allFiles[filePath].filter(...)`.

When you see the `forEach(... push(x))` it's a flag that it might be possible to `filter(...)` instead.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 30 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: Yes, I agree. Yesterday I was getting error on the type of `moduleSpecifiers` as with my logic, it should have been Promise<ModuleImport[]>[]. But, `createTsSourceFile(x)` is the one returning Promise. So, it makes sense a lot.

Comment (hansl) [inline addon/ng2/utilities/get-dependent-files.ts]: Yup. Think about your logic; you want to manipulate SourceFiles, not Promises, so it makes sense to resolve the promises as early as possible and just work with the values you actually want. It's less confusing for the reader as well because you don't go back and forth between using Arrays and using Promises. You use Promises once, then just do the actual work.

File: addon/ng2/utilities/get-dependent-files.ts (hunk 31 above)
Comment (ashoktamang) [inline addon/ng2/utilities/get-dependent-files.ts]: thanks for showing me the pattern. I will consider this approach in my implementation from now on. Refactoring it right now.

**Comment by hansl** [comment]
LGTM for me.  @alxhub do you want to review this?

File: addon/ng2/utilities/get-dependent-files.ts [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+import * as fs from 'fs';
+import * as ts from 'typescript';
+import * as glob from 'glob';
+import * as path from 'path';
+import * as denodeify from 'denodeify';
+
+import { Promise } from 'es6-promise';
+
+/**
+ * Declare an Interface where key is a string and value is either a string or a number.
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: I don't quite follow how this sentence relates to `ModuleImport`.

In general, a good comment should talk about what the object or interface _is_ instead of what's inside it.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+export function createTsSourceFile(fileName: string): Promise<ts.SourceFile> {
+  const readFile = denodeify(fs.readFile);
+  return readFile(fileName, 'utf8')
+    .then((contents: string) => {
+      return ts.createSourceFile(fileName, contents, ts.ScriptTarget.ES6, true);
+    });
+}
+
+/**
+ * @return traverses through ts.Node and returns an array of moduleSpecifiers
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Generally `@param` should precede `@return`, and you still need a comment describing the function itself.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+      return {
+        specifierText: moduleSpecifier.getText().slice(1, -1),
+        pos: moduleSpecifier.pos,
+        end: moduleSpecifier.end
+      };
+    });
+}
+
+/** 
+ * @return a boolean value after it searches for a barrel (index.ts by convention) in a given path
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Same as above - need a comment and reverse order.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 35]
```diff
@@ -0,0 +1,124 @@
...
+ * 
+ * @param filePath
+ * @param isDir boolean value to declare whether filePath is a directory or not. 
+ */
+export function hasIndexFile(filePath: string, isDir: boolean): Boolean {
+  let indexFile: string[];
+  if (isDir) {
+    indexFile = glob.sync(path.join(filePath, 'index.ts'), { nodir: true });
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Avoid synchronous operations.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * Return a map of all dependent file/s' path with their moduleSpecifier object
+ * (text, pos, end)
+ * 
+ * @param fileName file upon which other files depend 
+ * @param rootPath root of the project
+ * 
+ */
+export function getDependentFiles(fileName: string, rootPath: string): Promise<ModuleMap> {
+  const files = glob.sync(path.join(rootPath, '**/*.*.ts'), { nodir: true });
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Avoid sync operations.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 37]
```diff
@@ -0,0 +1,124 @@
...
+  return Promise.all(files.map(file => createTsSourceFile(file)))
+    .then((tsFiles: ts.SourceFile[]) => tsFiles.map(file => getImportClauses(file)))
+    .then((moduleSpecifiers: ModuleImport[][]) => {
+      let allFiles: ModuleMap = {};
+      for (let index = 0; index < files.length; index++) {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Prefer `Array.forEach` to an explicit loop.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 38]
```diff
@@ -0,0 +1,124 @@
...
+        allFiles[sourcePath] = moduleSpecifiers[index];
+      }
+      return allFiles;
+    })
+    .then((allFiles: ModuleMap) => {
+      let relevantFiles: ModuleMap = {};
+      Object.keys(allFiles).forEach(filePath => {
+        const tempModuleSpecifiers: ModuleImport[] = allFiles[filePath]
+          .filter(importClause => {
+            let modulePath = path.resolve(path.dirname(filePath), importClause.specifierText);
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: There are two types of imports in Typescript - relative and absolute.

A relative import begins with `.` - `./foo` or `../../bar`. An absolute import is something like `@angular/core`, which Typescript will handle the resolution of depending on `tsconfig.json`.

I think you should ignore non-relative imports, as attempting to resolve them this way would be incorrect behavior if a similarly named file happened to exist.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+    });
+}
+
+/** 
+ * @return a boolean value after it searches for a barrel (index.ts by convention) in a given path
+ * 
+ * @param filePath
+ * @param isDir boolean value to declare whether filePath is a directory or not. 
+ */
+export function hasIndexFile(filePath: string, isDir: boolean): Boolean {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: In general, boolean parameters are a bad idea in a method signature. They're hard to read and generally lead to poor APIs.

In this case, perhaps the caller should be responsible for providing only a directory name, instead of it being `hasIndexFile`'s responsibility to resolve the directory.

File: addon/ng2/utilities/get-dependent-files.ts [hunk 40]
```diff
@@ -0,0 +1,124 @@
...
+      return allFiles;
+    })
+    .then((allFiles: ModuleMap) => {
+      let relevantFiles: ModuleMap = {};
+      Object.keys(allFiles).forEach(filePath => {
+        const tempModuleSpecifiers: ModuleImport[] = allFiles[filePath]
+          .filter(importClause => {
+            let modulePath = path.resolve(path.dirname(filePath), importClause.specifierText);
+            let resolvedFileName = path.resolve(fileName);
+            if (hasIndexFile(path.resolve(fileName), false)) {
```
Comment (alxhub) [inline addon/ng2/utilities/get-dependent-files.ts]: Just handle the no index case for now. The index case is considerably more complicated.
